Fossilize: Fix multiple crash and bad exports

ishitatsuyuki opened the pull request:
## Description

Each change address a different issue that resulted in fossilize exports that either crashes or produce unintended results.

The code are tested, but I'm still not confident about the code structure:

- For `foz: Encode sampleMask as array`, is it a good idea to put the type conversion right inside `ConvertSDObjectToFossilizeJSON`?
- For `foz: Reconstruct specialization constants with correct layout`, should we instead make the SD reader export the spec constant buffers so we can export the pipeline byte-accurately?

ishitatsuyuki committed:
foz: Rename dynamic rendering colorAttachmentFormats

ishitatsuyuki committed:
foz: Reconstruct specialization constants with correct layout

ishitatsuyuki committed:
foz: Encode sampleMask as array

baldurk commented on qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp:
```diff
@@ -4363,7 +4375,17 @@ QVariant VulkanPipelineStateViewer::ConvertSDObjectToFossilizeJSON(const SDObjec
+          {
+            map[key] = v;
+          }
+        }
```
This should be handled externally. Look at how `AddFossilizeNexts` works in-place on the generated maps to patch them, a similar thing can be done to modify `sampleMask`.

baldurk commented on renderdoc/api/replay/vk_pipestate.h:
```diff
@@ -596,6 +596,15 @@ value.
 )");
   bytebuf specializationData;
+
+  // NOTE: This mirrors patchData->specIDs.
```
Minor nitpick - remove this comment as this header is public-facing so shouldn't reference internal implementation details.

baldurk commented on qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp:
```diff
@@ -4216,20 +4216,29 @@ QByteArray VulkanPipelineStateViewer::ReconstructSpecializationData(const VKPipe
+    int32_t idx = sh.specializationIds.indexOf(constantId);
+    if(idx == -1)
+    {
+      qWarning("Couldn't find original data for spec ID %u, substituting with zero", constantId);
```
This warning can be dropped I think - I found it's common for applications to over-specify constant IDs at specialisation time that aren't used in the shader, which is entirely reasonable if you have fixed constant ID allocation and don't always use them all. You have the resize above so the buffer is big enough, and the value in them is immaterial.

baldurk commented on qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp:
```diff
@@ -4216,20 +4216,29 @@ QByteArray VulkanPipelineStateViewer::ReconstructSpecializationData(const VKPipe
+      qWarning("Couldn't find original data for spec ID %u, substituting with zero", constantId);
+      continue;
+    }
+    size_t srcByteOffset = idx * sizeof(uint64_t);
```
This packing shouldn't be assumed here, the offset in the shader's `specializationData` in the pipeline state is given by the `byteOffset` member of the shader constant. The array is parallel to the `specializationIds` you've added, so you can look up the corresponding constant directly in the reflection data. That way if the packing changes, the data is still found.

baldurk commented on renderdoc/api/replay/vk_pipestate.h:
```diff
@@ -596,6 +596,15 @@ value.
+  // NOTE: This mirrors patchData->specIDs.
+  DOCUMENT(R"(Specialization constant indices for each entry in specializationData, representing
+the specialization constant layout. Element at `specializationData[i*8]` refers to specialization
+constant `specializationIds[i]`.
```
It feels a bit wrong needing to 'leak' this information to the UI as it shouldn't be needed, but indeed I can't think of a cleaner solution while still generating the fossilize DB in the UI. Since constant IDs are no longer used directly as an offset (unfortunately necessary as some applications use the full 32-bit range) they can no longer be implicit so some correlation is needed.

Storing the spec ID on every constant in the reflection info like the default value is too much (and I am strongly considering removing that default value from there in future refactors). Saving and sharing the 'unmodified' specialisation buffer just for this purpose is overkill and likely to cause confusion as to which buffer to use since they're arranged differently.

The only thing I'd suggest is rewording the comment here. The data is still the same either way, but this array should be described in terms of its relationship to the list of shader constants in the specialisation constant `ConstantBlock`. In the same order, giving their API constant IDs. This then allows correlation between a constant ID and an offset by going via the shader constant for its `byteOffset` as noted above. Avoid making a direct comparison between the IDs and where they're stored in case the packing needs to change again in future.

baldurk reviewed (changes requested):
Thanks for this change. I think the solution is the best we can do but as in the comments above it should be described slightly differently.

ishitatsuyuki committed:
spirv_reflect: Remove unnecessary spec sort

By construction the variables are pushed in order from smaller offset to
larger offsets. The sort only confuses the reader, so remove it.

ishitatsuyuki commented:
I'm sorry for the delay, but I've pushed a revision to address the comments. I've also added a change to remove the sorting of spec const reflection variables — in the current code they are sorted by construction and an explicit code confuses the reader.

baldurk commented on qrenderdoc/Windows/PipelineState/VulkanPipelineStateViewer.cpp:
```diff
@@ -4368,6 +4395,12 @@ QVariant VulkanPipelineStateViewer::ConvertSDObjectToFossilizeJSON(const SDObjec
 
+      if(map.contains(lit("sampleMask")))
+      {
+        auto sampleMaskArray = QVariantList({map[lit("sampleMask")]});
```
Minor nit, please don't use `auto` except for things like lambdas or stl iterators. `QVariantList` here is preferred to `auto`.

baldurk reviewed (changes requested):
This looks great now, thanks. One minor nit on the code which I didn't comment before as I wasn't sure if it was going to stay given earlier feedback.

ishitatsuyuki commented:
Updated, thanks for the review!
